### ISSUE TITLE
Duplicate documents returned on alias scan

### ISSUE THREAD

**Issue by aewhite** [elastic/elasticsearch-hadoop#363]
I have an alias that points to several indices. After scanning this alias, I had ~50% more records than where actually indexed. After some research I think I found the issue.

Example of all indices pointed to by my alias. ABC are machines and the numbers represent shards...

```
               A                  B                 C
index-x        0                  1                 2         
index-y        0                  2                 1
```

When ES-Haddop resolves it's input splits, it does generate two separate splits for index-x-0 and index-y-0, but when the actual scan happen, it appears to use the preference API to select the shard. That *would* work if it sent the resolved index along too but it sends the alias name instead. Thus, the readers scan table-x-0 AND table-y-0. 

This issue was greatly multiplied in my environment because many indices had the same shard number on the same node. 

__edit:__ changed wording from "tables" to "indices"

**Comment by costin** [comment]
Can you please confirm if I understand correctly - the alias is resolved properly to its respective shards (in this case shards for `x` and `y`) however when it sends the requests (in this case two splits for `x` and two for `y`) it still uses the alias which means when reading `x`, the `y` shards locally residing will be used as well and the same thing will happen for `y`.
Correct?

**Comment by aewhite** [comment]
I think you got it. I am pretty sure that if you just persisted that index name with the shard input split and use the resolved index names directly, the bug would be solved.

**Comment by costin** [comment]
@aewhite Hi, I've managed to reproduce the bug and pushed a fix for it in master and 2.x. Can you please try the 2.1 dev build, it is already up in the Maven repo.

Thanks!

**Comment by aewhite** [comment]
The current 2.1 dev build fails with the following exception...

```
2015-01-28 01:40:07,020 FATAL org.apache.hadoop.mapred.Child: Error running child : java.lang.IncompatibleClassChangeError: Found interface org.apache.hadoop.mapreduce.TaskAttemptContext, but class was expected
	at org.elasticsearch.hadoop.mr.EsInputFormat.createRecordReader(EsInputFormat.java:440)
	at org.elasticsearch.hadoop.mr.EsInputFormat.createRecordReader(EsInputFormat.java:64)
	at org.apache.hadoop.mapred.MapTask.runNewMapper(MapTask.java:644)
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:330)
	at org.apache.hadoop.mapred.Child$4.run(Child.java:268)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:415)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1408)
	at org.apache.hadoop.mapred.Child.main(Child.java:262)
```

**Comment by costin** [comment]
Sorry about that - it's fixed now in master and a new build has been pushed.

**Comment by aewhite** [comment]
Better but I'm still seeing more documents exported than exist in ES by about 10%. I'll start digging but it might take a while given our dataset size and ES/cluster setup.

**Comment by aewhite** [comment]
I think it's picking up replicas where their shard number overlaps with a primary shard of the same number from another index.

**Comment by costin** [comment]
@aewhite Hi, I've pushed a fix in master and the associated version in Maven. It took significantly longer than expected since a improvement in elasticsearch is needed (see elasticsearch/elasticsearch#9483). That might be available in 1.5 but until then, the solution is a best effort - when dealing with aliases or multiple indices (`foo,bar,tar` - and any of them can be an alias), es-hadoop tries to see if there's any combination of nodes where shards (with the same id) do not overlap. If there is one, it will use it - if not it will disable parallelism and do only one request.
With the fix in Elasticsearch, the parallelism would not be reduced but rather increased (since it will result in a task/worker/query for each shard in each index. If the alias points to 3 indices, that would result in 3x more queries than today).

Can you please try the latest dev version and report back? Thanks!

**Comment by costin** [comment]
@aewhite Any update? did you manage to try out the latest snapshot?

**Comment by aewhite** [comment]
Sorry, I have not and it's going to be difficult to test now due to topology changes and other priorities. We now have fewer smaller indices and fewer shards in general. 

Since we needed a workaround for a production release, we ended up submitting parallel jobs to Hadoop, one per physical index. 

That said, I did review your changes and they look promising but unfortunately I'm not in a position to replicate the original problem easily.

**Comment by rjhdeboer** [comment]
This fix has worked for me.

Before this fix, the difference between the number of results from a regular ES query and the RDD started to deviate exponentially the more indices I queried (e.g. for 1 index the numbers were the same, for 2 indices the RDD containing exactly twice as many results, for 3 indices there were more than triple the number of results).

After the fix, the number of results were the same no matter how many indices I queried.
